### ISSUE TITLE
Take slices that maintain gradient information

### ISSUE THREAD

**Issue by YLGH** [pytorch/pytorch#89857]
### 🚀 The feature, motivation and pitch

I have a use case where I want to take slices of a tensor, but then have the gradients of those slices be consistent with that of the original tensor slice: e.g.

`bag = nn.EmbeddingBag(10, 4, device="cuda", mode="sum")

bag_weight_1 = bag.weight[0:5]
bag_weight_1.retain_grad()

indices = torch.tensor([0,1], device="cuda").int()
offsets = torch.tensor([0], device="cuda").int()
out = bag(indices, offsets)
out.sum().backward()

print(bag.weight, bag.weight.grad)
print(bag_weight_1.grad)`

The last line prints none, but Ideally i'd want it to be a slice of bag.weight.grad.

Is there something like this that is possible today? It definitely feels a bit weird as the bag_weight_1 parameter/tensor doesn't in the forward call or the autograd

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by ngimel** [comment]
As you say, since `bag_weight_1` doesn't participate in forward computation, you shouldn't expect it to have gradient.

**Comment by albanD** [comment]
I'm afraid this is expected behavior indeed.

One thing you can do is to do it the other way around: have bag_weight_1 be used in a differentiable way to generate the weight:
```python
import torch
from torch import nn

bag = nn.EmbeddingBag(10, 4, mode="sum")

bag_weight_1 = bag.weight[0:5].detach().requires_grad_()
new_weight = torch.cat([bag_weight_1, bag.weight[5:].detach()])
new_weight.retain_grad()
del bag.weight; bag.weight = new_weight

indices = torch.tensor([0,1]).int()
offsets = torch.tensor([0]).int()
out = bag(indices, offsets)
out.sum().backward()

print(bag.weight, bag.weight.grad)
print(bag_weight_1.grad)
```

Note though that bag.weight is NOT a nn.Parameter here anymore.
If you do want to learn bag_weight_1, you can make it a Parameter before the cat and set it on the bag Module.

If you just wanted to access the gradients, doing `bag.weight.grad[0:5]` is the simplest thing for you to do.

**Comment by YLGH** [comment]
Hey folks, thanks for the replies, 

yeah i think it makes a lot of sense that slices don't have the grads attached,

We ended up doing something like this instead:

https://github.com/pytorch/torchrec/blob/main/torchrec/distributed/composable/table_batched_embedding_slice.py#L15

we had to override the grad/grad_fn properties, so that we can override it with bag.weight.grad[...],

it "works", but curious to hear your feedback on it/if this could cause issues down the line,

Thanks + happy new years! 

@albanD  @ngimel
